sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotHideBaseClassMethods.cs

```diff
@@ -0,0 +1,180 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotHideBaseClassMethods : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4019";
+        private const string MessageFormat = "Remove or rename that method because it hides '{0}'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var classDeclaration = c.Node as ClassDeclarationSyntax;
+                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+
+                    if (classSymbol == null || classDeclaration.Identifier.IsMissing)
+                    {
+                        return;
+                    }
+
+                    var issueFinder = new IssueFinder(classSymbol, c.SemanticModel);
+
+                    classDeclaration
+                        .Members
+                        .OfType<MemberDeclarationSyntax>()
+                        .Select(issueFinder.FindIssue)
+                        .WhereNotNull()
+                        .ToList()
+                        .ForEach(c.ReportDiagnostic);
+                },
+                SyntaxKind.ClassDeclaration);
+        }
+
+        private class IssueFinder
+        {
+            private readonly IList<IMethodSymbol> allBaseClassMethods;
+            private readonly SemanticModel semanticModel;
+
+            public IssueFinder(INamedTypeSymbol classTypeSymbol, SemanticModel semanticModel)
+            {
+                this.semanticModel = semanticModel;
+                allBaseClassMethods = GetAllBaseMethods(classTypeSymbol).ToList();
+            }
+
+            private static IEnumerable<IMethodSymbol> GetAllBaseMethods(INamedTypeSymbol classTypeSymbol)
+            {
+                while (classTypeSymbol?.BaseType != null)
```

Comment:
This could probably be written without two loops
```
return classTypeSymbol.BaseType.GetSelfAndBaseTypes().SelectMany(t => t.GetMembers()).Where(...);
```